Cudnn RNN bindings.

abergeron opened the pull request:
This survived some very light testing (it can run forwards and backwards without crashing).

I did not compare this implementation against a "known-good" one yet, but I plan to do that later. 

In the meantime, if anybody has feedback on the code, especially the first commits which add new functionality, that would be great.

nouiz commented:
travis and jenkins fail.

nouiz commented:
I think the travis problem is that it use python 2.6. I'll make another PR to drop it as we discussed

nouiz commented:
https://github.com/Theano/Theano/pull/4924 when merged, you can rebase this PR on top of it.

nouiz commented:
```
/home/travis/build/Theano/Theano/theano/gpuarray/dnn.py:1797:13: E126 continuation line over-indented for hanging indent

/home/travis/build/Theano/Theano/theano/gpuarray/dnn.py:1829:27: E261 at least two spaces before inline comment

/home/travis/build/Theano/Theano/theano/gpuarray/dnn.py:1830:28: E261 at least two spaces before inline comment

/home/travis/build/Theano/Theano/theano/gpuarray/dnn.py:1834:32: E261 at least two spaces before inline comment

/home/travis/build/Theano/Theano/theano/gpuarray/dnn.py:1840:18: E261 at least two spaces before inline comment

/home/travis/build/Theano/Theano/theano/gpuarray/dnn.py:1841:19: E261 at least two spaces before inline comment

/home/travis/build/Theano/Theano/theano/gpuarray/dnn.py:1845:23: E261 at least two spaces before inline comment
```

nouiz commented:
In Python 3.3, there is a bunch of errors like:

```
======================================================================

ERROR: test_in_update (theano.compile.tests.test_function.TestFunctionIn)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/build/Theano/Theano/theano/compile/tests/test_function.py", line 88, in test_in_update

    mode='FAST_RUN')

  File "/home/travis/build/Theano/Theano/theano/compile/function.py", line 326, in function

    output_keys=output_keys)

  File "/home/travis/build/Theano/Theano/theano/compile/pfunc.py", line 484, in pfunc

    output_keys=output_keys)

  File "/home/travis/build/Theano/Theano/theano/compile/function_module.py", line 1789, in orig_function

    defaults)

  File "/home/travis/build/Theano/Theano/theano/compile/function_module.py", line 1653, in create

    input_storage=input_storage_lists, storage_map=storage_map)

  File "/home/travis/build/Theano/Theano/theano/gof/link.py", line 703, in make_thunk

    storage_map=storage_map)[:3]

  File "/home/travis/build/Theano/Theano/theano/gof/vm.py", line 1120, in make_all

    self.updated_vars,

  File "/home/travis/build/Theano/Theano/theano/gof/vm.py", line 958, in make_vm

    ivar.container)

AttributeError: 'TensorVariable' object has no attribute 'container'
```

nouiz commented on theano/gof/op.py:
```diff
@@ -694,6 +694,9 @@ def __ne__(self, other):
     # Python implementation #
     #########################
 
+    def grad2(self, inputs, outputs, output_grads):
```
As discussed a few weeks ago. I don't like the idea to have a method named grad2.

I would prefer that at the call place we have a try... except. We try with the extra args, if we get an AttributeError, then we call the old version. This would be better from a User that make Op perspective, but it have others problems.

Do you have others solution? A completly new names? I think this method don't really implement the gradient, but something close. Could this be the occasion to also please people very good in math by reconciling this other problem and not have a function ending with 2?

nouiz replied:
This issue won't block this PR as it isn't documented for now. But if we merge it like this, we need an issue about it and hopefully we should find a good fix not too long. I would like a fix before 0.9 to don't release new interface that we could change later.

abergeron replied:
I'm all for a new, better name.  I just don't have any good ideas.

nouiz replied:
I asked @slefrancois if he have ideas.

lamblin replied:
Would it be possible to either:
- use a metaclass in Op so that would automatically redefine the `grad` method so it accepts (and ignore) `outputs`, or
- use `inspect` in `theano.gradient` to see what it accepts and pass that (rather than try/except)?

lamblin replied:
Otherwise, we could use `Lop` for the name. This is less easily understandable, but technically more correct than "grad".

abergeron replied:
We already have an `Lop`  (which is the mirror of `Rop`).

abergeron replied:
Also, it would be possible to do a number of yucky things to accept both signatures, (the simplest being a try..except in the theano.gradient method, like what was done to add the c_compiler argument), but I don't like complexifiying the call protocol like that.  It's better to have a single name with a single signature.

In the end if we really want to keep the grad() name, we could add a method called _grad() that would try both signatures on public method and returns what works.

slefrancois replied:
As for a name for grad2, I don't have a rigorous mathematical name. I can suggest "grad_chain", since it's essentially the gradient of the op times the previous outputs from the chain rule.

abergeron replied:
I don't mind renaming, I just want a consensus.

lamblin replied:
Since `Op.Lop` does not actually exist, and that `theano.gradient.Lop` actually calls `theano.grad`, I would favor that one.
But I have no real objection against `grad_chain`, except it is less descriptive.

nouiz commented on theano/gof/type.py:
```diff
@@ -603,26 +633,32 @@ class CDataType(Type):
-    import ctypes
-    if PY3:
-        _cdata_type = ctypes.py_object.from_address(
-            ctypes.addressof(ctypes.pythonapi.PyCapsule_Type)).value
```
As we are droping python 2.6, I think we can clean up the code to replace PyCObject with PyCapsule usage everywhere. Can you confirm this? I make a note about this in the issue about dropping python 2.6

abergeron replied:
This is what this chunk is doing (in part).

nouiz commented on theano/gof/type.py:
```diff
@@ -603,26 +633,32 @@ class CDataType(Type):
-    del ctypes
-
-    def __init__(self, ctype, freefunc=None):
+    def __init__(self, ctype, freefunc=None, headers=None, header_dirs=None,
```
This probably need updates to the doc.

nouiz replied:
I think we should use the same name as compile_str and same order to be consistent:

    def compile_str(module_name, src_code, location=None,
                    include_dirs=None, lib_dirs=None, libs=None,
                    preargs=None, py_module=True, hide_symbols=True):

abergeron replied:
I use the same names as the methods you have to define on a type (minus the `c_` prefix).

abergeron replied:
Also, yes, the docs could be improved to reflect the new arguments, I'll do that after I get the tests working.

nouiz replied:
ok. Sad that we use different name for different stuff.

nouiz commented on theano/gof/type.py:
```diff
@@ -633,10 +669,22 @@ def __hash__(self):
 
+    def _get_func(self):
+        from theano.scalar import get_scalar_type
+        import theano
```
move at least the "import theano" at the top. I'm not sure if the other can be, but if possible do it.

abergeron replied:
I really wanted to avoid circular imports and this is the definition of Type so I don't think we can move the other one, but for "import theano" sure.

nouiz commented on theano/gof/type.py:
```diff
@@ -633,10 +669,22 @@ def __hash__(self):
+
+        if self._fn is None:
+            v = get_scalar_type('int64')()
+            self._fn = theano.function([v], _make_cdata(self)(v))
```
add the parameter profile=False to remove it from the user profile.

abergeron replied:
ok

abergeron commented:
I dropped the commit about pinned shared variables because it turns out that the final code doesn't need it and I don't want to debug it (It's the source of the "container" problem).

Apart from that, I've made flake8 fixes and addressed some small comments.

nouiz commented on theano/gof/type.py:
```diff
@@ -710,8 +742,20 @@ def c_cleanup(self, name, sub):
+
     def c_code_cache_version(self):
-        return (2, self.ctype, self.freefunc)
+        return (3,)
```
Why did you remove the other attribute and not add the new ones instead?

abergeron replied:
Because those should not be part of the cache key.  They are attributes of the type and are checked in \__eq__ and \__hash__.

nouiz replied:
right! It is not mandatory, but I think the new parameter should be added in \__eq__ and \__hash__. Shoud we use \__props__ for this class instead?

abergeron replied:
I've added an implementation of __props__ for all theano classes (at the object2 level) and used it here with the missing items.

nouiz commented on theano/gpuarray/type.py:
```diff
@@ -20,6 +20,7 @@
     pygpu = None
 
 _context_reg = {}
+_props_map = {}
```
add in a comment an example of props. props is an overloaded name with many meaning.

Or better, add docstring to the new user methods from this file.

abergeron replied:
ok

nouiz replied:
still needed to do

nouiz commented on theano/gpuarray/dnn.py:
```diff
@@ -39,6 +40,44 @@
+            # Update these names when new versions of cudnn are supported.
+            lib_name = ctypes.util.find_library('cudnn64_5.dll')
+            if lib_name is None:
+                lib_name = ctypes.util.find_library('cudnn64_4.dll')
```
Can you make a list and iterate on it? It will make adding new version much easier and it would remove duplicate code (1 line for now, but more later)

abergeron replied:
We only ever support one version back so the list shouldn't grow in size.

nouiz commented on theano/gpuarray/dnn.py:
```diff
@@ -39,6 +40,44 @@
+            if lib_name is None:
+                lib_name = ctypes.util.find_library('cudnn64_4.dll')
+        if lib_name is None:
+            raise RuntimeError('Could not find cudnn library')
```
We probably need a better error message, in case people use never cudnn version on windows.

nouiz commented on theano/gpuarray/__init__.py:
```diff
@@ -87,19 +87,24 @@ def init_dev(dev, name=None):
+                          (ctx_props['cudnn_version'], name), file=sys.stderr)
+                ctx_props['cudnn_handle'] = dnn._make_handle(context)
+            except Exception:
+                pass
```
This need at least a warning that cudnn isn't found and won't be used/available.

abergeron replied:
There wasn't any warnings in that case before and I don't see why we should start now.

nouiz replied:
In the old back-end, we tell it when cudnn isn't there. There is big consequence of this. We want to lower the number of user questions on the mailing list. It will help them find the problem if we print it. Otherwise, the problems/errors won't always be super clear.

nouiz commented on theano/gpuarray/dnn.py:
```diff
@@ -153,20 +198,31 @@ class DnnBase(COp):
+            return get_prop(self.dnn_context(node), 'cudnn_handle_param')
+        except KeyError:
+            pass
+        ptr = get_prop(self.dnn_context(node), 'cudnn_handle').value
```
Will we have raised a better error before if cudnn isn't available? If not sure, just add a try except and tell that cudnn is needed.

abergeron replied:
Maybe.  Not if you attempt to use cudnn ops directly, bypassing the optimizers.

nouiz commented on theano/gpuarray/dnn_base.c:
```diff
@@ -103,44 +103,7 @@ c_set_filter(PyGpuArrayObject *var, cudnnFilterDescriptor_t desc) {
-    cuda_exit(PARAMS->ctx);
-    FAIL;
-  }
-  if ((err = cudnnSetStream(APPLY_SPECIFIC(_handle),
```
I don't see any new code that would call setstream. So it will probably end up in the default stream instead of the stream we want.

abergeron replied:
It will end up in the NULL stream which is ok now that libgpuarray is ready for that.

nouiz commented on theano/gpuarray/dnn_base.c:
```diff
@@ -103,44 +103,7 @@ c_set_filter(PyGpuArrayObject *var, cudnnFilterDescriptor_t desc) {
-#section cleanup_code_struct
-
-cuda_enter(ctx->ctx);
-cudnnDestroy(APPLY_SPECIFIC(_handle));
```
It mean anything we put in the props like the cudnnhandle is never destroyed. I don't see problems with that. Just wanted to make sure you saw this and also don't expect problems with this.

abergeron replied:
It is destroyed at python shutdown.  And it not so much a big deal since we don't create a handle per op anymore.

nouiz commented on theano/gpuarray/dnn_base.c:
```diff
@@ -88,6 +92,13 @@ c_set_filter(PyGpuArrayObject *var, cudnnFilterDescriptor_t desc) {
     dims[i] = PyGpuArray_DIM(var, i);
   }
 
+  /* Filters can't be less than 3d so we pad */
```
Can you add in this comment and the previous like this which current version of cudnn have this limit. In case it change in the futur, it is useful.

abergeron replied:
I have no idea when this limit started.

nouiz replied:
I mean, add that at current cudnn v5, there is this limit. So if we end up in the futur where the current and the previous don't have this limit and we still see this code, we now easily we can remove it. Also, if we drop cudnn v5, it make the clean up faster and make sure we won't forget to check that.

nouiz commented on theano/gpuarray/dnn_base.c:
```diff
@@ -114,6 +126,17 @@ c_set_filter(PyGpuArrayObject *var, cudnnFilterDescriptor_t desc) {
   return 0;
 }
 
+static int c_make_filter(PyGpuArrayObject *var, cudnnFilterDescriptor_t *desc) {
```
This can probably be used to clean up code. If that is the case, it would be great to include it. Quick to do and make a big difference in readability.

nouiz commented:
I think you removed the commit that add support for pinned shared var. Is this wanted?

nouiz commented on theano/gpuarray/dnn.py:
```diff
@@ -1630,6 +1632,325 @@ def make_node(self, x, dy, scale, x_mean, x_invstd):
+        assert context_name == self.context_name
+        # This is a dirty hack to pass the context because params is
+        # occupied by the cudnn handle
+        context = gpu_context_type.make_constant(get_context(context_name))
```
instead of this ugly hack. Can you make a GpuArrayConstant() on the right context and pass this as the 3rd input? It would be less hacky and would like like what the other cudnn op that need the context do except that this input is only used for this. It will then allow to cache this c code. I really hate not being able to cache the c code...

abergeron replied:
That would be even more of a hack.  The proper solution is to implement a struct-like type and use it to pass multiple values in params.

nouiz replied:
I think we should not merge this until we enable the caching of this. We already compile too much. Using a constant isn't very hacky and cause less problems. You use the inputs, not the value, but its context.

I agree the full solution is better, but I think it would be another PR in itself. So I would go with the constant inputs here.

nouiz commented on theano/gpuarray/dnn.py:
```diff
@@ -1630,6 +1632,325 @@ def make_node(self, x, dy, scale, x_mean, x_invstd):
+
+def _make_dropout_desc(dropout, seed, context_name):
+    desc, states = theano.function([], _DropoutDescriptor(context_name)(
+        dropout, seed, context_name))()
```
Cache the function building and pass the parameter profile=True. Should we also pass the mode Mode(optimizer=None)?

abergeron replied:
I can't cache the function since it would cache the values for dropout and seed.

Also since this is a 2-node function I don't think we will have any speed problems with it.

nouiz commented on theano/gpuarray/dnn.py:
```diff
@@ -1630,6 +1632,325 @@ def make_node(self, x, dy, scale, x_mean, x_invstd):
+
+def _make_rnn_desc(hidden_size, num_layers, ddesc, rnn_mode,
+                   input_mode, direction_mode, dtype, context_name):
+    desc = theano.function([], _RNNDescriptor(context_name)(
```
idem. cache the function and pass profile=False. Maybe Mode(optimizer=None)

abergeron replied:
also idem.

nouiz replied:
if you can't cache, do the others.

nouiz commented on theano/gpuarray/dnn.py:
```diff
@@ -1630,6 +1632,325 @@ def make_node(self, x, dy, scale, x_mean, x_invstd):
+    def dnn_context(self, node):
+        return self.context_name
+
+    def do_constant_folding(self, node):
```
Should we make the default do_constant_folding be False if any of the inputs or outputs are CDataType?

abergeron replied:
Maybe.

nouiz commented on theano/gpuarray/dnn.py:
```diff
@@ -1630,6 +1632,325 @@ def make_node(self, x, dy, scale, x_mean, x_invstd):
+def _get_param_size(desc, input_size, dtype, context_name):
+    typecode = gpuarray.dtype_to_typecode(dtype)
+    return theano.function([], _RNNParamSize(context_name)(
+            desc, input_size, typecode))()
```
idem, cache, profile=False, maybe Mode(optimizer=None)

nouiz commented on theano/gpuarray/dnn.py:
```diff
@@ -1630,6 +1632,325 @@ def make_node(self, x, dy, scale, x_mean, x_invstd):
+class GpuDnnRNNOp(DnnBase):
+    __props__ = ()
+    _cop_numi = 5
+    _cop_numo = 4
```
Can you confirm that this is only needed as this op have a variable number of inputs and that the COp don't support that?

I don't like the name, _cop_num_inputs would be better. Why not add an extra Constant input that is NULL to the op instead of this hack? Can an CDataType instance be NULL?

It seem to work, but I don't like it. I don't have other suggestion through...

abergeron replied:
I've added support for optional inputs/outputs for COp and this is what you need to add.  I'll rename the values.

nouiz commented on theano/gpuarray/dnn.py:
```diff
@@ -1630,6 +1632,325 @@ def make_node(self, x, dy, scale, x_mean, x_invstd):
+        if cx is None:
+            context_name = infer_context_name(w, x, hx, cx)
+        else:
+            context_name = infer_context_name(w, x, hx)
```
this if seem strange. if cx is None, I would not pass it to infer_context_name. In fact, I would inverse the content of both condition. If I'm wrong, it would need a comment.

abergeron replied:
No you're right and this is reversed.

nouiz replied:
Still not fixed.

abergeron commented:
Yes I removed the code for pinned variables since I don't need it in the end and I don't have time to debug it right now.  I've kept the commit in one of my branches somewhere for later.

abergeron commented:
The only thing broken with the tests now should be that python 2.6 doesn't work anymore.

nouiz commented:
This PR fix gh-4257

abergeron commented:
This will need https://github.com/Theano/Theano/pull/4951 and https://github.com/Theano/Theano/pull/4968 to be merged first and then a rebase.

abergeron commented:
I'll fix conflicts once the tests pass locally.

For now I have the forward mode passing the tests as well as the gradient with regards to inputs.

As for the gradient on weights, it returning super high values (comparing to the reference) and I'll need to find out why.

abergeron commented:
I finally fixed all the problems and rebased, so now this should be ready to merge.

nouiz commented:
some travis error:

```
/home/travis/build/Theano/Theano/theano/configparser.py:92:1: E302 expected 2 blank lines, found 1

/home/travis/build/Theano/Theano/theano/gpuarray/dnn.py:1938:9: E301 expected 1 blank line, found 0

/home/travis/build/Theano/Theano/theano/gpuarray/dnn.py:2003:36: E226 missing whitespace around arithmetic operator

/home/travis/build/Theano/Theano/theano/gpuarray/tests/rnn_support.py:1:1: FI10 __future__ import "division" missing
```

nouiz commented:
More travis error here: https://travis-ci.org/Theano/Theano/jobs/161476634

nouiz commented on theano/gpuarray/dnn.py:
```diff
@@ -1582,6 +1640,559 @@ def make_node(self, x, dy, scale, x_mean, x_invstd):
+    def make_node(self, w, desc, layer, isize, typecode):
+        w = as_gpuarray_variable(w, infer_context_name(w))
+        layer = as_scalar(layer).astype('int32')
+        isize = as_tensor_variable(isize).astype('uint64')
```
assert ndim on isize and w.

nouiz commented on theano/gpuarray/dnn.py:
```diff
@@ -1582,6 +1640,559 @@ def make_node(self, x, dy, scale, x_mean, x_invstd):
+        elif direction_mode == 'unidirectional':
+            self.num_dirs = 1
+        else:
+            raise ValueError('direction_mode')
```
Tell that the value is invalid and show what was given.

nouiz commented on theano/gpuarray/dnn.py:
```diff
@@ -1582,6 +1640,559 @@ def make_node(self, x, dy, scale, x_mean, x_invstd):
+            inputs.append(cx)
+
+        _3d = GpuArrayType(dtype=x.dtype, broadcastable=(False, False, False),
+                           context_name=context_name)
```
I think we should do at some point as in the old back-end and the CPU, define fmatrix, dmatrix, matrix (that take a dtype param), instead of making new type everytimes. Can be later.

lamblin replied:
Agreed.
Actually, I would make a case for the default `matrix()` to instantiate a `GpuArrayType` if `device` is not `cpu`.
But that is a different issue.

nouiz commented on theano/gpuarray/dnn.py:
```diff
@@ -1582,6 +1640,559 @@ def make_node(self, x, dy, scale, x_mean, x_invstd):
+        _, dy, dhy = output_grads[:3]
+        dcy = output_grads[3] if len(output_grads) == 4 else None
+        # If both dy and dhy are disconnected, then this will error
+        # out, but it is indeed an error.
```
Can you explain why this is happening? This isn't something usual.

abergeron replied:
Yes, indeed.

nouiz commented on theano/gpuarray/dnn.py:
```diff
@@ -1582,6 +1640,559 @@ def make_node(self, x, dy, scale, x_mean, x_invstd):
+                 input_mode='linear', direction_mode='unidirectional',
+                 dropout=0.0, dropout_seed=4242, context_name=None):
+        """
+        dtype: data type of computation
```
This seem to be the user interface. Add some text.

Do we want to start to add that in Theano? Should this be moved to the test? Should we compare what framework interface exists to try to make them match?

@lamblin what do you think?

lamblin replied:
I'm not sure what the issue is. Is it the class a typical user would call?

abergeron replied:
Yes it is the user interface.  We don't really have a choice to add something like this here since we need a central object to hold the related parameters during graph creation.

I consider this a one-off and not the start of a pattern.  In that sense I would not try to make it match anything nor would I try to ensure future endeavours will match this.  If we ever get a cudnnGetRNNDescriptor() then we could probably do away with this and should.

lamblin replied:
There should be a warning that even though `dropout` and `dropout_seed` are parameter, only `dropout=0` is supported and the seed should not matter then.
Or maybe remove them from the interface if they are not really supported.

abergeron replied:
I'll remove them.

nouiz commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -1095,3 +1097,78 @@ def test_batchnorm_inference():
+    Y = T.tensor3('Y')
+    Y.tag.test_value = numpy.zeros((timesteps, batch_size, hidden_dim), dtype=theano.config.floatX)
+    h0 = T.tensor3('h0')
+    h0.tag.test_value = numpy.zeros((depth, batch_size, hidden_dim), dtype=theano.config.floatX)
```
remove test_value. I don't think they help the test now.

nouiz commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -1095,3 +1097,78 @@ def test_batchnorm_inference():
+    last_layer = WrapperLayer(X.dimshuffle(1, 0, 2))
+    last_dim = input_dim
+    for i in range(depth):
+        gru = GRU(last_dim, hidden_dim, last_layer, s0=h0[i, :, :])
```
It would be great to extend the tests to more parameter, event just to make sure they compile.

Not sure how priority it is.

abergeron commented on theano/configparser.py:
```diff
@@ -89,38 +89,43 @@ def config_files_from_theanorc():
+        try:
+            for k, v in iteritems(self.confs):
+                v.__set__(None, self.new_vals[k])
+        except:
```
We want to run during everything.  It shouldn't have an effect outside of the `with` statement.

lamblin replied:
I agree with @abergeron on that one.

abergeron commented on theano/gof/type.py:
```diff
@@ -678,8 +673,10 @@ def _get_func(self):
         from theano.scalar import get_scalar_type
```
ok

abergeron commented on theano/gpuarray/dnn.py:
```diff
@@ -1582,6 +1640,559 @@ def make_node(self, x, dy, scale, x_mean, x_invstd):
+        return self.context_name
+
+    def do_constant_folding(self, node):
+        return False
```
Otherwise it tries to do constant folding and fails.

lamblin commented on theano/gpuarray/dnn.py:
```diff
@@ -55,11 +55,12 @@ def _dnn_lib():
+                    break
         if lib_name is None:
-            raise RuntimeError('Could not find cudnn library')
+            raise RuntimeError('Could not find cudnn library (maybe your are using a )')
```
a what?

abergeron commented:
I did another rebase and added a test for lstm mode and fixed the issues of the gradient of that.

nouiz commented:
there is conflict in this branch.

abergeron commented:
Rebased again.

lamblin commented on theano/gpuarray/dnn.py:
```diff
@@ -1586,6 +1645,597 @@ def make_node(self, x, dy, scale, x_mean, x_invstd, epsilon=1e-4):
+        # that's ok as in that case we really are disconnected.
+        if isinstance(dy.type, DisconnectedType):
+            dy = as_gpuarray_variable(dhy[-1],
+                                      context_name=dhy.type.context_name)
```
So, if `dy` is `DisconnectedType`, I think we should pass zeros of the right size.
If  `dhy` is disconnected as well, then I think the `Lop` method would not be called at all.

abergeron replied:
* `dcy` could be connected (for LSTM), while `dy` and `dhy` are disconnected.
 * `y == hy[-1]`.  I verified this.  So it follows that `dy == dhy[-1]`.
 * We get an uninformative message even without the fix if dhy is disconnected, unless we use the special handling I am doing.
 * I can try leaving dy NULL if it's disconnected, but the documentation seemed to imply that it was a required parameter.
   * If that doesn't work I'll add a test for dhy connected and dy disconnected.

lamblin replied:
- I meant, if all outputs are disconnected, `Lop` should not be called.
- If that is true even with multiple layers, it means that the shapes given in lines 2064 and following do not make sense. Either `y` should have a "numLayers" somewhere, or `hy` should have "seqLength".
  - Even then, just because the values are equal does not mean we should pass the gradient twice.

I still think the right value to pass to `dy` if disconnected is `zeros_like(y)`.

abergeron replied:
I'll change it to zeros_like then.

lamblin commented on theano/gpuarray/dnn.py:
```diff
@@ -1586,6 +1645,597 @@ def make_node(self, x, dy, scale, x_mean, x_invstd, epsilon=1e-4):
+          unidirectional: The network operates recurrently from the
+                          first input to the last.
+
+          bidirectional: The operates from first to last then from last to first and concatenates the results at each layer.
```
"the network operates"

abergeron commented on theano/gpuarray/dnn.py:
```diff
@@ -1586,6 +1645,597 @@ def make_node(self, x, dy, scale, x_mean, x_invstd, epsilon=1e-4):
+
+
+def _make_dropout_desc(dropout, seed, context_name):
+    desc, states = theano.function(
```
Better to disable them than to get bitten later.

lamblin replied:
I'm not sure what would bite us the harder: not disabling them, or disabling them and preventing users from using test values in that part of the code.

abergeron commented on theano/gpuarray/dnn.py:
```diff
@@ -1586,6 +1645,597 @@ def make_node(self, x, dy, scale, x_mean, x_invstd, epsilon=1e-4):
+    %(fail)s
+  }
+  off = (intptr_t)o - (intptr_t)w;
+  assert(off >= 0 && "bias");
```
I am (reasonably) confident it will never trigger, but if it does, then it will be much easier to debug than if it wasn't there.

Also, I disagree that asserts are hard to debug.  They print out a clear error message and are super easy to catch with gdb.

abergeron replied:
I'm ok with that.

lamblin commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -1095,3 +1097,150 @@ def test_batchnorm_inference():
+    ref_fn, ref_grad_fn = funcs(last_layer.output(),
+                                model.get_params())
+    cudnn_fn, cudnn_grad_fn = funcs(rnnb.apply(params_cudnn, X, h0)[0],
+                                    [params_cudnn])
```
We would also need a test where the output (that is used for the cost) depends on the hidden state at the last step (hy) only, and on a mixture of hy and y.
This is the only way of checking that `dhy` and `dy` work correctly together.

lamblin commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -1095,3 +1097,150 @@ def test_batchnorm_inference():
+    ref_fn, ref_grad_fn = funcs(last_layer.output(),
+                                model.get_params())
+    cudnn_fn, cudnn_grad_fn = funcs(rnnb.apply(params_cudnn, X, h0, c0)[0],
+                                    [params_cudnn])
```
Same here, we should test a cost that depends on `cy` as well.

abergeron commented:
I don't have any reference code for bidirectional rnns and I would probably waste a lot of time trying to come up with some.

Also, I'm not trying to write regression tests for cudnn, just for the bindings and the bidirectional case is basically a larger hidden_size for everything, so if the current tests pass, that should work too.

The rest of the comments should be addressed and I also rebased.

lamblin commented:
For bidirectional RNNs, I was mostly worried about the memory layout of the initial states, final states, and outputs.
But indeed, the reference code would be harder. We can leave that to future work.

lamblin commented:
I'll merge because I think it is ready, but we should make sure it is documented online.
It can be a simple `audoc` on the main class, but it should be searchable from Google and the built-in search box.
